### PR TITLE
v4 Rewrite

### PR DESCRIPTION
**Pull request by englercj**
v4 of Resource Loader, now written in Typescript and with pluggable loading strategies.

### CODE REVIEW

**Comment by englercj** [comment]
@bigtimebuddy If you have some time, I'd appreciate a review. I know it is a big change.

**Comment by englercj** [comment]
@GoodBoyDigital this is the v4 rewrite I was mentioning to you that had custom load strategies support.

**Comment by bigtimebuddy** [comment]
Wow, awesome work @englercj. I’ll look through it a little closer tomorrow, but in general the rollup, docs and package scripts looked really solid to me.

**Comment by GoodBoyDigital** [comment]
weeee! nice one @englercj !

File: README.md [hunk 1]
```diff
@@ -62,9 +78,13 @@ That will output the built distributables to `./dist`.
...
 - Opera 12.1+
 
-## Upgrading to v2
+## Upgrading to v4
 
-- No more events, all signals now
-- No more isJson, isXml, etc. Now use `res.type === Resource.TYPE.JSON`, etc.
-- Removed `before` (in favor of `pre`) and `after` (in favor of `use`).
-- If a middleware adds more resources, it *must* pass in the parent resource in options for `.add()`.
+- Before middleware has been removed, so no more `pre` function.
```
Comment (bigtimebuddy) [inline README.md]: Why no more pre middleware support?

Comment (bigtimebuddy) [reply README.md]: There is one use case that I find `pre` very helpful which is preprocessing a URL. Two practical examples: one is processing language token for localizing content (eg, `my-file-%LOCALE%.jpg`), the other is something I setup with pixi-sound to sniff for a format type (eg, `my-file.{omg,mp3}`) I find this incredibly convenient to  handle at the loader level. I hope you may reconsider deprecating this feature even if the caching middleware isn’t right for it.

Comment (englercj) [reply README.md]: Interesting, I hadn't ever heard of anyone using this feature so I decided to just remove the complexity of having it. If people are actually using it I don't mind putting it back in.

Comment (bigtimebuddy) [reply README.md]: Overall, I agree with you about caching and removing that feature. But definitely having a way to preprocess URLs is very handy. Could also be used to implement resolution fetching e.g., `file@2x.png`

Comment (englercj) [reply README.md]: Would it also be useful to allow the user to set a url resolve callback that will be called to resolve urls? Maybe as a replacement for the `baseUrl` concept is the ability to just assign url resolvers that can do that logic and more?

Comment (bigtimebuddy) [reply README.md]: That makes more sense than pre. Customs URL resolvers is good and accomplished the same thing.

Comment (englercj) [reply README.md]: Sorry I haven't gotten back to this. I worked this weekend, and next weekend is Blizzcon so may be a couple weeks before I can get back here and do the right thing. Sorry!

Comment (englercj) [reply README.md]: Alright, I've added a new `urlResolver` property which is a function that gets to modify the url before use.

**Pull request merged**